builtin/read.c

```diff
@@ -55,8 +51,36 @@ int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
 		mrsh_buffer_append_char(&buf, (char)c);
 	}
 	mrsh_buffer_append_char(&buf, '\0');
+	
+	struct mrsh_word_string *str = mrsh_word_string_create(mrsh_buffer_steal(&buf), false);
+	
+	struct mrsh_array fields = {0};
+	split_fields(&fields, &str->word, mrsh_env_get(state, "IFS", NULL));
 
-	mrsh_env_set(state, argv[optind], buf.data, MRSH_VAR_ATTRIB_NONE);
-	mrsh_buffer_finish(&buf);
+	if (fields.len <= argc - optind) {
+		for (size_t i = 0; i < fields.len; ++i) {
+			mrsh_env_set(state, argv[optind + i], (char*)fields.data[i], MRSH_VAR_ATTRIB_NONE);
+		}
+		for (size_t i = fields.len; i < argc - optind; ++i) {
+			mrsh_env_set(state, argv[optind + i], "", MRSH_VAR_ATTRIB_NONE);
+		}
+	} else {
+		for (int i = 0; i < argc - optind - 1; ++i) {
+			mrsh_env_set(state, argv[optind + i], (char*)fields.data[i], MRSH_VAR_ATTRIB_NONE);
+		}
+		struct mrsh_buffer buf_last = {0};
+		for (int i = argc - optind - 1; i < fields.len; ++i) {
+			char *field = (char*)fields.data[i];
+			mrsh_buffer_append(&buf_last, field, strlen(field));
+			if (i != fields.len - 1) {
+				// TODO add the field delimiter rather than space (bash and dash always use spaces)
+				mrsh_buffer_append_char(&buf_last, ' ');
+			}
+		}
+		mrsh_env_set(state, argv[argc - 1], mrsh_buffer_steal(&buf_last), MRSH_VAR_ATTRIB_NONE);
+	}
+	
+	mrsh_word_destroy(&str->word);
+	mrsh_array_finish(&fields);
```

Comment:
Each field is a `char *` that needs to be free'd. We need to iterate over `fields` to free each item.